### ISSUE TITLE
curly: "multi-or-nest" option.

### ISSUE THREAD

**Issue by andyearnshaw** [eslint/eslint#1806]
I've long been a fan of the (seemingly unpopular) curly style that allows omitting the braces for a single statement.  However, I like to enforce braces if inside another block, for example:

```javascript
// Acceptable
for (let i = 0; i < l; i++) {
    if (a[i])
        fn();
}

if (cond) {
    while (cond2) {
        for (cond3 in foo)
            bar();
    }
}
baz();

// Should be a warning
for (let i = 0; i < l; i++)
    if (a[i])
        fn();

if (cond)
    while (cond2)
        for (cond3 in foo)
            bar();
baz();
```

Generally, I find braces to be unnecessary clutter, but I like to prevent a large de-indent when continuing on after nested blocks.  Currently, I can't enforce this style with eslint.  The first form gives a warning regardless of whether the `curly` is enabled or disabled.

Any chance this could be implemented?

**Comment by nzakas** [comment]
What does ESLint output for this code now?

**Comment by andyearnshaw** [comment]
Currently, if I have `curly: [2, "multi"]` set I get something like this:

> 41:12  error  Unnecessary { after 'for' condition    curly

**Comment by nzakas** [comment]
Sorry for not being clear, I'd like to see the code plus the ESLint output so I can see what it's flagging and why.

**Comment by andyearnshaw** [comment]
This is a sample of some code I have:

```javascript
for (let i = 0, m = modules.length; i < m; i++) {
    if (modules[i].domains.some(matchHost))
        return makeAPIRequest(modules[i], urlParts).then(resolve, reject);
}
```

This is the error eslint gives with `curly: [2, "multi"]`:

```
   41:12  error  Unnecessary { after 'for' condition    curly
```

**Comment by nzakas** [comment]
As this has been open for a few months with no progress, we will close this unless someone comes forward to champion this change.

**Comment by BYK** [comment]
I have bitten by this with a code like following:

    if (something)
        if (some_other_condition)
            do_X();
    else
            do_Z();

As you can see the `else` belongs to the inner `if` block and the only way to manage this to add braces if you want the `else` block to belong the outer `if`. When you do that, this rule complains about unnecessary braces. More over, no indenting rule caught this error. This is especially jarring when you are coming over from Python :)

I'd like to try working on this if no one else is gonna do it.

**Comment by nzakas** [comment]
@BYK feel free

**Comment by inikulin** [comment]
@BYK Any progress on this?

**Comment by BYK** [comment]
@inikulin - not yet. Turned a bit more complicated than I expected. You can take over if you are willing to work on it but I plan to get to this by this or next week.

**Comment by inikulin** [comment]
I'm working on it.

**Comment by BYK** [comment]
Alright then.
